compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -320,29 +324,32 @@ object Implicits:
     }
 
     /** The implicit references that are eligible for type `tp`. */
-    def eligible(tp: Type): List[Candidate] =
-      if (tp.hash == NotCached)
-        Stats.record(i"compute eligible not cached ${tp.getClass}")
-        Stats.record(i"compute eligible not cached")
-        computeEligible(tp)
-      else {
-        val eligibles = eligibleCache.lookup(tp)
-        if (eligibles != null) {
+    def eligible(tp: Type, enableUnsafeNulls: Boolean = false): List[Candidate] =
+      def searchWithCache(cache: EqHashMap[Type, List[Candidate]]) = {
+        val eligibles = cache.lookup(tp)
+        if eligibles != null then
           Stats.record("cached eligible")
           eligibles
-        }
-        else if (irefCtx eq NoContext) Nil
-        else {
+        else if irefCtx eq NoContext then Nil
+        else
           Stats.record(i"compute eligible cached")
-          val result = computeEligible(tp)
-          eligibleCache(tp) = result
+          val result = computeEligible(tp, enableUnsafeNulls)
+          cache(tp) = result
           result
-        }
       }
+      if tp.hash == NotCached then
+        Stats.record(i"compute eligible not cached ${tp.getClass}")
+        Stats.record(i"compute eligible not cached")
+        computeEligible(tp, enableUnsafeNulls)
+      else if enableUnsafeNulls then
+        searchWithCache(eligibleCacheUnsafeNulls)
+      else
+        searchWithCache(eligibleCache)
 
-    private def computeEligible(tp: Type): List[Candidate] = /*>|>*/ trace(i"computeEligible $tp in $refs%, %", implicitsDetailed) /*<|<*/ {
+    private def computeEligible(tp: Type, enableUnsafeNulls: Boolean): List[Candidate] = /*>|>*/ trace(i"computeEligible $tp in $refs%, %", implicitsDetailed) /*<|<*/ {
       if (monitored) record(s"check eligible refs in irefCtx", refs.length)
-      val ownEligible = filterMatching(tp)
+      val c = if enableUnsafeNulls then ctx.addMode(Mode.UnsafeNullConversion) else ctx
```

Comment:
The `ctx` in `computeEligible` is the `initctx` from `ImplicitRefs`, which is different from the typing context. Hence, it may not contain the `unsafeNulls` Mode.